server/src/main/java/org/elasticsearch/transport/MessageHeader.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.transport;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.core.Tuple;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+
+public abstract class MessageHeader {
```

Comment:
These class names need a rethink

Reply:
I don't really understand why this PR splits the `Header` into multiple classes.

I recognize we might need a specialized path to parse a handshake `Header`. But that is just a method, not multiple classes with super classes, etc.

Reply:
This is primarily aimed at the future work to separate out handshake entirely.

By creating separate classes, it immediately declares the places where it can be both types (handshake vs data) or it should just be processing just data packets.

It is also an easy way to declare & use the two different types of versions (TransportVersion vs int) whilst maintaining the common code used for both types.